test/kuka_kr6_test.cpp

```diff
@@ -0,0 +1,113 @@
+#include <array>
+#include <gtest/gtest.h>
+#include <random>
+
+#include "opw_kinematics/opw_kinematics.h"
+#include "opw_kinematics/opw_parameters_examples.h" // for makeIrb2400_10<double>()
+#include "opw_kinematics/opw_utilities.h"           // for optional checking
+
+template <typename T>
+using Transform = Eigen::Transform<T, 3, Eigen::Affine>;
+
+template <typename T>
+void comparePoses(const Transform<T> & Ta, const Transform<T> & Tb)
+{
+  using Matrix = Eigen::Matrix<T, 3, 3>;
+  using Vector = Eigen::Matrix<T, 3, 1>;
+
+  Matrix Ra = Ta.rotation(), Rb = Tb.rotation();
+  for (int i = 0; i < Ra.rows(); ++i)
+  {
+    for (int j = 0; j < Ra.cols(); ++j)
+    {
+      EXPECT_NEAR(Ra(i, j), Rb(i, j), 1e-6);
+    }
+  }
+
+  Vector pa = Ta.translation(), pb = Tb.translation();
+  EXPECT_NEAR(pa[0], pb[0], 1e-6);
```

Comment:
I think it would be interesting to make this epsilon value different for T = float and T = double. I was playing with the inverse code and dropped out all of the implicit conversions to doubles. This sped the code up by factor of two for floats (500k calls in 475k us, under 1us per solve!). The side effect is that 3 of those poses had like 1.2e-6 in error.

```
/home/jmeyer/ros/opw_ws/src/opw_kinematics/test/kuka_kr6_test.cpp:29: Failure
The difference between pa[0] and pb[0] is 1.1622905731201172e-06, which exceeds 1e-6, where
pa[0] evaluates to 0.33538031578063965,
pb[0] evaluates to 0.33538147807121277, and
1e-6 evaluates to 9.9999999999999995e-07.
/home/jmeyer/ros/opw_ws/src/opw_kinematics/test/kuka_kr6_test.cpp:29: Failure
The difference between pa[0] and pb[0] is 1.0728836059570312e-06, which exceeds 1e-6, where
pa[0] evaluates to 0.33538255095481873,
pb[0] evaluates to 0.33538147807121277, and
1e-6 evaluates to 9.9999999999999995e-07.
/home/jmeyer/ros/opw_ws/src/opw_kinematics/test/kuka_kr6_test.cpp:29: Failure
The difference between pa[0] and pb[0] is 1.1622905731201172e-06, which exceeds 1e-6, where
pa[0] evaluates to 0.33538031578063965,
pb[0] evaluates to 0.33538147807121277, and
1e-6 evaluates to 9.9999999999999995e-07.
/home/jmeyer/ros/opw_ws/src/opw_kinematics/test/kuka_kr6_test.cpp:29: Failure
The difference between pa[0] and pb[0] is 1.0430812835693359e-06, which exceeds 1e-6, where
pa[0] evaluates to 0.33538252115249634,
pb[0] evaluates to 0.33538147807121277, and
1e-6 evaluates to 9.9999999999999995e-07.
```

Given that a float can only hold about 7 sig-figs of data, I think this is okay. It's the trade off you make between computational speed, memory pressure, and precision.

Thanks for writing this.

Reply:
Wow, that's really fast, especially for code that works for so many robots, and more than accurate enough for my application I think. The `1e-6` is what was used in the MoveIt! tests I looked at for inspiration.

Maybe we can use `1e-5` for floats and something like `1e-10` for doubles (although that error stays under `1e-14` for this test when I run it for doubles.)

I will add the tolerance as a global const variable to the test. Or do I put this kind of settings somewhere else?